In [ ]:
!pip install -qU ragas==0.2.10
!pip install -qU langchain-community==0.3.14 langchain-openai==0.2.14 unstructured==0.16.12 langgraph==0.2.61 langchain-qdrant==0.2.0 nltk

### Keys

In [ ]:
import os
from getpass import getpass
os.environ["OPENAI_API_KEY"] = getpass("Please enter your OpenAI API key!")
os.environ["RAGAS_APP_TOKEN"] = getpass("Please enter your Ragas API key!")

Please enter your OpenAI API key!··········
Please enter your Ragas API key!··········


### SDG: Generating synthetic data

### Data Preparation

Download the webpages which we'll be using for our data today.

These webpages are from [Simon Willison's](https://simonwillison.net/) yearly "AI learnings".

- [2023 Blog](https://simonwillison.net/2023/Dec/31/ai-in-2023/)
- [2024 Blog](https://simonwillison.net/2024/Dec/31/llms-in-2024/)


In [ ]:
from langchain_community.document_loaders import DirectoryLoader
from ragas.llms import LangchainLLMWrapper
from ragas.embeddings import LangchainEmbeddingsWrapper
from langchain_openai import ChatOpenAI
from langchain_openai import OpenAIEmbeddings
from ragas.testset import TestsetGenerator

def dataPreparation():
  !mkdir data
  !curl https://simonwillison.net/2023/Dec/31/ai-in-2023/ -o data/2023_llms.html
  !curl https://simonwillison.net/2024/Dec/31/llms-in-2024/ -o data/2024_llms.html

  #Load the data into a familiar LangChain format using the DirectoryLoader
  path = "data/"
  loader = DirectoryLoader(path, glob="*.html")
  docs = loader.load()

  return docs

def dataGeneration(docs):
  generator_llm = LangchainLLMWrapper(ChatOpenAI(model="gpt-4o"))
  generator_embeddings = LangchainEmbeddingsWrapper(OpenAIEmbeddings())

  generator = TestsetGenerator(llm=generator_llm, embedding_model=generator_embeddings)
  dataset = generator.generate_with_langchain_docs(docs, testset_size=30)
  #print(dataset.to_pandas())
  print(dataset.upload())
  return dataset

In [ ]:
docs = dataPreparation()
dataset = dataGeneration(docs)

mkdir: cannot create directory ‘data’: File exists
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 31427    0 31427    0     0   140k      0 --:--:-- --:--:-- --:--:--  140k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 70286    0 70286    0     0   489k      0 --:--:-- --:--:-- --:--:--  490k


Applying HeadlinesExtractor:   0%|          | 0/2 [00:00<?, ?it/s]

Applying HeadlineSplitter:   0%|          | 0/2 [00:00<?, ?it/s]

Applying SummaryExtractor:   0%|          | 0/2 [00:00<?, ?it/s]

Applying CustomNodeFilter:   0%|          | 0/12 [00:00<?, ?it/s]

Applying [EmbeddingExtractor, ThemesExtractor, NERExtractor]:   0%|          | 0/26 [00:00<?, ?it/s]

Applying [CosineSimilarityBuilder, OverlapScoreBuilder]:   0%|          | 0/2 [00:00<?, ?it/s]

Generating personas:   0%|          | 0/2 [00:00<?, ?it/s]

Generating Scenarios:   0%|          | 0/3 [00:00<?, ?it/s]

Generating Samples:   0%|          | 0/30 [00:00<?, ?it/s]

Testset uploaded! View at https://app.ragas.io/dashboard/alignment/testset/5d76caaa-ee59-4c47-88d2-910fc4c62b88
https://app.ragas.io/dashboard/alignment/testset/5d76caaa-ee59-4c47-88d2-910fc4c62b88


## HW:


1.   LangGraph RAG with naive retrieval
2. Baseline Evaluation using RAGAS METRICS
3. Implement semantic chunking
4. Update LangGraph RAG to use semnatic chunking with naive retrieval
5. Run evaluations to compare and constract the results



### LangGraph RAG with naive retreival

#### Chunking and embedding

In [ ]:
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_openai import OpenAIEmbeddings

from langchain_openai import OpenAIEmbeddings
from langchain_qdrant import QdrantVectorStore
from qdrant_client import QdrantClient
from qdrant_client.http.models import Distance, VectorParams
import numpy as np
from nltk.tokenize import sent_tokenize
from langchain_core.documents import Document


class EmbeddingRetriever():

  def __init__(self, docs):
    self.docs = docs
    self.embeddings = OpenAIEmbeddings(model="text-embedding-3-small")
    self.split_documents = []

  def __createEmbeddingsAndRetriever(self, split_documents):
    client = QdrantClient(":memory:")

    client.create_collection(
        collection_name="ai_across_years",
        vectors_config=VectorParams(size=1536, distance=Distance.COSINE),
    )

    vector_store = QdrantVectorStore(
        client=client,
        collection_name="ai_across_years",
        embedding=self.embeddings,
    )

    vector_store.add_documents(documents=split_documents)
    retriever = vector_store.as_retriever(search_kwargs={"k": 5})

    return retriever

  # Perform semantic chunking.
  # For current version, chunk size is represented in number of lines.
  def __semanticChunking(self,min_chunk_size: 'int', max_chunk_size: 'int', threshold:'float'):
    split_documents = []
    chunk = ""
    chunk_size = 0
    for doc in self.docs:
      for sentence in sent_tokenize(doc.page_content):
        if chunk_size<min_chunk_size:
          chunk+=sentence
          chunk_size+=1
          continue

        chunk_embedding = self.embeddings.embed_query(chunk)
        sentence_embedding = self.embeddings.embed_query(sentence)
        similarity_score = self.__cosine_similarity(chunk_embedding,sentence_embedding)

        if similarity_score >= threshold:
          chunk += sentence
          chunk_size+=1

        if chunk_size >= max_chunk_size or similarity_score <threshold:
          split_documents.append(Document(page_content=chunk, metadata=doc.metadata))
          chunk = ""
          chunk_size = 0

        #print(f"Processing chunk:{len(split_documents)}")

    return split_documents

  def __cosine_similarity(self,v1,v2):
    dot_product = np.dot(v1, v2)
    norm_v1 = np.linalg.norm(v1)
    norm_v2 = np.linalg.norm(v2)
    return dot_product / (norm_v1 * norm_v2)


  def __chunking(self, stratergy):
    if stratergy["algorithm"] == "recursiveCharacterTextSplitter":
      #chunk_overlap parameter specifically controls how much overlap there is between consecutive chunks
      text_splitter = RecursiveCharacterTextSplitter(chunk_size=stratergy["chunk_size"],
                                                     chunk_overlap=stratergy["chunk_overlap"])
      self.split_documents = text_splitter.split_documents(self.docs)

    if stratergy["algorithm"] == "semanticChunking":
      self.split_documents = self.__semanticChunking(stratergy["min_chunk_size"],
                                                    stratergy["max_chunk_size"],
                                                     stratergy["threshold"])

    return self.split_documents

  def getRetriever(self, stratergy):
    split_documents = self.__chunking(stratergy)
    print(f"split_documents length:{len(split_documents)}")
    return self.__createEmbeddingsAndRetriever(split_documents)


#### R - Retrieval

In [ ]:
embedding_retriever_util = EmbeddingRetriever(docs)
retriever = embedding_retriever_util.getRetriever({
    "algorithm":"recursiveCharacterTextSplitter",
    "chunk_size":1000,
    "chunk_overlap":200})

def retrieve(state):
  retrieved_docs = retriever.invoke(state["question"])
  return {"context" : retrieved_docs}

split_documents length:74


#### A - Augument

In [ ]:
from langchain.prompts import ChatPromptTemplate

RAG_PROMPT = """\
You are a helpful assistant who answers questions based on provided context. You must only use the provided context, and cannot use your own knowledge.

### Question
{question}

### Context
{context}
"""

rag_prompt = ChatPromptTemplate.from_template(RAG_PROMPT)

#### G - Generation

Using gpt-4o-mini

In [ ]:
from langchain_openai import ChatOpenAI

llm = ChatOpenAI(model="gpt-4o-mini")

def generate(state):
  docs_content = "\n\n".join(doc.page_content for doc in state["context"])
  messages = rag_prompt.format_messages(question=state["question"], context=docs_content)
  response = llm.invoke(messages)
  return {"response" : response.content}

#### LangGraph

In [ ]:
from langgraph.graph import START, StateGraph
from typing_extensions import List, TypedDict
from langchain_core.documents import Document

class State(TypedDict):
  question: str
  context: List[Document]
  response: str

graph_builder = StateGraph(State).add_sequence([retrieve, generate])
graph_builder.add_edge(START, "retrieve")
graph = graph_builder.compile()


In [ ]:
response = graph.invoke({"question" : "How are LLM agents useful?"})
response["response"]

'LLM agents, or Large Language Model agents, are useful in several ways:\n\n1. **Ease of Building**: They are relatively easy to construct, requiring only a few hundred lines of Python code, provided that adequate training data is available. This accessibility allows more people to experiment with and create LLMs, broadening their potential applications.\n\n2. **Running on Personal Devices**: Recent advancements have made it possible to run LLMs on personal computers, making the technology more accessible to individuals and smaller organizations without the need for expensive servers.\n\n3. **Code Generation**: LLMs have shown particular effectiveness in generating code. Their ability to write and execute code, particularly with tools like the ChatGPT Code Interpreter, allows users to verify the correctness of generated code. This capability is less susceptible to the issue of hallucination compared to other applications of LLMs.\n\n4. **Potential for AI Agents**: There is excitement a

### Evaluation

In [ ]:
for test_row in dataset:
  response = graph.invoke({"question" : test_row.eval_sample.user_input})
  test_row.eval_sample.response = response["response"]
  test_row.eval_sample.retrieved_contexts = [context.page_content for context in response["context"]]

In [ ]:
dataset.to_pandas()

user_input  \
0   Wht did OpenAI kno that the rest of us didnt a...   
1   What is the description of the butterfly photo...   
2     What advancements were made with GPT-4 in 2023?   
3   Wht iz the role of AGI in the context of evals...   
4   Wht is the role of OpenAI in the context of en...   
5   Wht role does Meta play in the development of ...   
6           What stuff we learned about LLMs in 2023?   
7                           What Llama stuff you got?   
8   What role does Stability AI play in the develo...   
9   How does the use of JavaScript relate to the c...   
10  What are the ethical implications of the envir...   
11  How do the challenges of gullibility in AI age...   
12  Why people criticize LLMs and how the criticis...   
13  How have advancements in LLM efficiency impact...   
14  How have advancements in AI efficiency impacte...   
15  How have model efficiency improvements impacte...   
16  How have advancements in AI models impacted th...   
17  How has the criticism and skepticism of LLMs e...   
18  How has the environmental impact of AI models ...   
19  What advancements in AI and Large Language Mod...   
20  How does the pricing of Claude 3 compare to ot...   
21  How has the development of GPT-4o and its pric...   
22  What were the key advancements in Large Langua...   
23  How has the development of Claude 3 and its pr...   
24  What advancements in AI and Large Language Mod...   
25  How do the pricing and capabilities of GPT-4o ...   
26  How has Claude been utilized in the developmen...   
27  What were the key developments and challenges ...   
28  What are the challenges and potential solution...   
29  How did the advancements in 2024 lead to the b...   

                                   retrieved_contexts  \
0   [If you can gather the right data, and afford ...   
1   [How good are those descriptions? Here’s what ...   
2   [This is a huge advantage for open over closed...   
3   [A lot of people are excited about AI agents—a...   
4   [I think this means that, as individual users,...   
5   [So training an LLM still isn’t something a ho...   
6   [This is Things we learned about LLMs in 2024 ...   
7   [Meta’s Llama 3.2 models deserve a special men...   
8   [I wrote about how Large language models are h...   
9   [It’s still astonishing to me how effective th...   
10  [Since then, almost every major LLM (and most ...   
11  [Gullibility is the biggest unsolved problem\n...   
12  [LLMs need better criticism\n\nA lot of people...   
13  [Since then, almost every major LLM (and most ...   
14  [OpenAI are not the only game in town here. Go...   
15  [I can now run a GPT-4 class model on my lapto...   
16  [I think this means that, as individual users,...   
17  [LLMs need better criticism\n\nA lot of people...   
18  [Law is not ethics. Is it OK to train models o...   
19  [This is a huge advantage for open over closed...   
20  [I think this means that, as individual users,...   
21  [This is a huge advantage for open over closed...   
22  [Simon Willison’s Weblog\n\nSubscribe\n\nThing...   
23  [So training an LLM still isn’t something a ho...   
24  [This is a huge advantage for open over closed...   
25  [Today $30/mTok gets you OpenAI’s most expensi...   
26  [Getting back to models that beat GPT-4: Anthr...   
27  [Simon Willison’s Weblog\n\nSubscribe\n\nStuff...   
28  [Simon Willison’s Weblog\n\nSubscribe\n\nStuff...   
29  [Training a GPT-4 beating model was a huge dea...   

                                   reference_contexts  \
0   [Prompt driven app generation is a commodity a...   
1   [gets you OpenAI’s most expensive model, o1. G...   
2   [feed with the model and talk about what you c...   
3   [dependent on AGI itself. A model that’s robus...   
4   [that. DeepSeek v3 is a huge 685B parameter mo...   
5   [Another common technique is to use larger mod...   
6   [Simon Willison’s Weblog Subscribe Things we l...   
7   [Those of us who understand this stuff have a .

In [ ]:
from ragas import EvaluationDataset

evaluation_dataset = EvaluationDataset.from_pandas(dataset.to_pandas())


In [ ]:
#judge model
from ragas import evaluate
from ragas.llms import LangchainLLMWrapper

evaluator_llm = LangchainLLMWrapper(ChatOpenAI(model="gpt-4o"))

In [ ]:
from ragas.metrics import LLMContextRecall, Faithfulness, FactualCorrectness, ResponseRelevancy, ContextEntityRecall, NoiseSensitivity
from ragas import evaluate, RunConfig

custom_run_config = RunConfig(timeout=360)

result = evaluate(
    dataset=evaluation_dataset,
    metrics=[LLMContextRecall(), Faithfulness(), FactualCorrectness(), ResponseRelevancy(), ContextEntityRecall(), NoiseSensitivity()],
    llm=evaluator_llm,
    run_config=custom_run_config
)
result

Evaluating:   0%|          | 0/180 [00:00<?, ?it/s]

ERROR:ragas.executor:Exception raised in Job[65]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[71]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[101]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[107]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[125]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[131]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[137]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[149]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[167]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[173]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[179]: TimeoutError()


{'context_recall': 0.7118, 'faithfulness': 0.8062, 'factual_correctness': 0.4040, 'answer_relevancy': 0.9505, 'context_entity_recall': 0.4061, 'noise_sensitivity_relevant': 0.2766}

### Update chunking stratergy and re-evaluate

In [ ]:
semantic_chunking_retriever_util = EmbeddingRetriever(docs)
semantic_chunking_retriever = semantic_chunking_retriever_util.getRetriever({
    "algorithm":"semanticChunking",
    "min_chunk_size":1,
    "max_chunk_size":100,
    "threshold":0.70})

def semantic_chunking_retrieve(state):
  retrieved_docs = semantic_chunking_retriever.invoke(state["question"])
  return {"context" : retrieved_docs}

split_documents length:211


In [ ]:
semantic_chunking_retriever_util.split_documents[0:5]

[Document(metadata={'source': 'data/2024_llms.html'}, page_content='Simon Willison’s Weblog\n\nSubscribe\n\nThings we learned about LLMs in 2024\n\n31st December 2024\n\nA lot has happened in the world of Large Language Models over the course of 2024.'),
 Document(metadata={'source': 'data/2024_llms.html'}, page_content='This is a sequel to my review of 2023.'),
 Document(metadata={'source': 'data/2024_llms.html'}, page_content='The environmental impact got better\n\nThe environmental impact got much, much worse\n\nThe year of slop\n\nSynthetic training data works great\n\nLLMs somehow got even harder to use\n\nKnowledge is incredibly unevenly distributed\n\nLLMs need better criticism\n\nEverything tagged “llms” on my blog in 2024\n\nThe GPT-4 barrier was comprehensively broken\n\nIn my December 2023 review I wrote about how We don’t yet know how to build GPT-4—OpenAI’s best model was almost a year old at that point, yet no other AI lab had produced anything better.'),
 Document(metada

In [ ]:
class State(TypedDict):
  question: str
  context: List[Document]
  response: str

new_graph_builder = StateGraph(State).add_sequence([semantic_chunking_retrieve, generate])
new_graph_builder.add_edge(START, "semantic_chunking_retrieve")
new_graph = new_graph_builder.compile()

In [ ]:
semantic_chunking_response = new_graph.invoke({"question" : "How are LLM agents useful?"})
semantic_chunking_response["response"]

"LLM agents are useful because they can perform a wide range of tasks, often surprising both users and the people who trained them with their capabilities. They have the potential to accomplish things that may not have been anticipated. However, it's important to recognize that LLMs can also exhibit significant limitations, as they may believe any information provided to them. Therefore, the key to effectively utilizing LLMs lies in understanding how to navigate their inherent unreliability while harnessing their powerful features."

In [ ]:
import time

for test_row in dataset:
  semantic_chunking_response = graph.invoke({"question" : test_row.eval_sample.user_input})
  test_row.eval_sample.response = semantic_chunking_response["response"]
  test_row.eval_sample.retrieved_contexts = [context.page_content for context in semantic_chunking_response["context"]]
  time.sleep(2) # To try to avoid rate limiting.

In [ ]:
semantic_chunking_result = evaluate(
    dataset=evaluation_dataset,
    metrics=[LLMContextRecall(), Faithfulness(), FactualCorrectness(), ResponseRelevancy(), ContextEntityRecall(), NoiseSensitivity()],
    llm=evaluator_llm,
    run_config=custom_run_config
)
semantic_chunking_result

Evaluating:   0%|          | 0/180 [00:00<?, ?it/s]

ERROR:ragas.executor:Exception raised in Job[17]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[65]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[71]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[83]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[101]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[107]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[119]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[125]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[131]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[137]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[143]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[149]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[155]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[161]: TimeoutError()
ERROR:ragas.executor:Exception raised in Job[167]: TimeoutError()
ERROR:ragas.ex

{'context_recall': 0.6990, 'faithfulness': 0.8419, 'factual_correctness': 0.4120, 'answer_relevancy': 0.9521, 'context_entity_recall': 0.4267, 'noise_sensitivity_relevant': 0.3027}

In [ ]:
semantic_chunking_result

{'context_recall': 0.6990, 'faithfulness': 0.8419, 'factual_correctness': 0.4120, 'answer_relevancy': 0.9521, 'context_entity_recall': 0.4267, 'noise_sensitivity_relevant': 0.3027}

In [ ]:
result

{'context_recall': 0.7118, 'faithfulness': 0.8062, 'factual_correctness': 0.4040, 'answer_relevancy': 0.9505, 'context_entity_recall': 0.4061, 'noise_sensitivity_relevant': 0.2766}

### Compare and Contrast the results:

| Metric | Semantic Chunking | Recursive Character Text Splitter | Comparison |
|---|---|---|---|
| `context_recall` | 0.6990 | 0.7118 | `Recursive Character Text Splitter` slightly higher (+1.28%) - indicates slightly better retrieval of relevant context. |
| `faithfulness` | 0.8419 | 0.8062 | `Semantic Chunking` significantly higher (+3.57%) - suggests responses are more grounded in the provided context. |
| `factual_correctness` | 0.4120 | 0.4040 | `Semantic Chunking` slightly higher (+0.8%) - indicates slightly more factually accurate answers. |
| `answer_relevancy` | 0.9521 | 0.9505 | `Semantic Chunking` slightly higher (+0.16%) - suggests slightly more relevant answers to the questions. |
| `context_entity_recall` | 0.4267 | 0.4061 | `Semantic Chunking` higher (+2.06%) - indicates better recall of important entities from the context. |
| `noise_sensitivity_relevant` | 0.3027 | 0.2766 | `Semantic Chunking` higher (+2.61%) - suggests slightly better robustness to irrelevant information. |

**Summary**:

Recursive Character Text Splitter performs negligibly better only in terms of retrieving relevant context (context_recall).
Semantic Chunking performs better across all other crucial metrics, most notably faithfulness, suggesting higher accuracy and grounding in provided context. It also shows marginal improvements in factual_correctness, answer_relevancy, context_entity_recall, and noise_sensitivity_relevant. <br/>

**Overall:**

Based on these results, Semantic Chunking appears to produce slightly more reliable and accurate responses compared to Recursive Character Text Splitter for this specific dataset and RAG pipeline configuration. Although Recursive Character Text Splitter retrieves marginally more relevant context, sSemantic Chunking demonstrates a better ability to utilize that context effectively and provide faithful and factually sound answers.

**Note**: The results are specific to the various parameters selected for given run of test. Ideally we should run multiple evaluation runs with different settings to identify better option